### ISSUE TITLE
Weighted subjects dataset to improve severe class imbalance

### ISSUE THREAD

**Issue by mattwarkentin** [fepegar/torchio#633]
**🚀 Feature**
A new type of `Dataset` that inherits from `SubjectsDataset` but can over/undersample classes to achieve better class balance. Perhaps something like: `WeightedSubjectsDataset` or something...

**Motivation**

In the medical world (but perhaps more generally), it may be the case that you may have much fewer positive cases compared to negative cases. A random sampler will maintain that disproportionality during training and so many steps may pass without the network seeing a single positive case.

**Pitch**

This new class would be "aware" of which key in the `Subject` corresponds to the label and the user could perhaps provide a dictionary with weights (similar to `tio.OneOf`) that indicates how they would like the label balance.

For example, if the real class balance is 900 negative and 100 positive (i.e. 90% vs. 10%), the user could do something like:
```py
weights = {'negative': 0.5, 'positive': 0.5}
tio.WeightedSubjectsDataset(subjects_list, transform=transforms, weights=weights)

## Sequence instead of dict would mean equal prob. by default
weights = {'negative', 'positive'}
tio.WeightedSubjectsDataset(subjects_list, transform=transforms, weights=weights)
```

There are a lot more implementation details that would need to be figured out, but does this potential feature make sense? Do you see the value? In your work, do you come up against severe class imbalance and how do you usually handle it within the `PyTorch`/`torchio` world?

**Comment by mattwarkentin** [comment]
Perhaps this is something that should be handled by the `DataLoader()`, using a `torch.utils.data.sampler.WeightedRandomSampler` but not sure if this plays nicely with the `SubjecsDataset` or not. Also it's my understanding that it might not play nicely with `PyTorch Lightning`...
